File: notebooklm_client.py
import asyncio
import json
import random
import re
import time
import os
import logging
import base64
from typing import Optional, Dict, Any, List
from playwright.async_api import async_playwright, Page, BrowserContext

logger = logging.getLogger(__name__)

# --- CONFIGURATION ---
BASE_URL = "https://notebooklm.google.com"
RPC_ENDPOINT = f"{BASE_URL}/_/LabsTailwindUi/data/batchexecute"

# RPC IDs
RPC_CREATE_NOTEBOOK = "CCqFvf"
RPC_ADD_SOURCE = "izAoDd"
RPC_GENERATE_INFOGRAPHIC = "R7cb6c"
RPC_LIST_ARTIFACTS = "gArtLc"
RPC_DELETE_NOTEBOOK = "f61S6e"

class NotebookLMClient:
    def __init__(self, headless: bool = True):
        self.headless = headless
        self.browser = None
        self.context = None
        self.page = None
        self.session_tokens = {"at": None, "bl": None, "fsid": None}
        self.cookies = []

    async def start(self):
        """Starts the browser and authenticates."""
        playwright = await async_playwright().start()
        # Use a persistent context to save login state
        base_dir = os.path.dirname(os.path.abspath(__file__))
        user_data_dir = os.path.join(base_dir, "user_data")
        self.context = await playwright.chromium.launch_persistent_context(
            user_data_dir,
            headless=self.headless,
            args=[
                "--disable-blink-features=AutomationControlled",
                "--ignore-certificate-errors", 
                "--no-sandbox",
                "--disable-setuid-sandbox"
            ]
        )
        self.page = self.context.pages[0] if self.context.pages else await self.context.new_page()
        
        await self._refresh_tokens()

    async def stop(self):
        if self.context:
            await self.context.close()

    async def _refresh_tokens(self) -> bool:
        """Navigates to NotebookLM and scrapes tokens."""
        logger.info("[NotebookLM] 🔄 Navigating to scrape tokens...")
        await self.page.goto(BASE_URL)
        
        # simple check for login
        if "accounts.google.com" in self.page.url:
            logger.warning("[NotebookLM] ⚠️ Login required! Please log in within the opened browser window.")
            # If we are headless, we can't login easily.
            # Ideally, the user should run once headed to login.
            if self.headless:
                 raise Exception("Authentication required. Please run with headless=False first to login.")
            
            # Wait for user to login (wait until we are back on notebooklm)
            await self.page.wait_for_url("https://notebooklm.google.com/**", timeout=0) # wait indefinitely
            logger.info("[NotebookLM] Login detected.")

        content = await self.page.content()
        
        at_match = re.search(r'"SNlM0e":"([^"]+)"', content)
        bl_match = re.search(r'"(boq_labs-tailwind-[^"]+)"', content)
        fsid_match = re.search(r'"FdrFJe":"([^"]+)"', content)

        if not at_match or not bl_match:
            # Fallback: maybe we need to wait a bit more for SPA to load?
            await asyncio.sleep(2)
            content = await self.page.content()
            at_match = re.search(r'"SNlM0e":"([^"]+)"', content)
            bl_match = re.search(r'"(boq_labs-tailwind-[^"]+)"', content)
            fsid_match = re.search(r'"FdrFJe":"([^"]+)"', content)

        if not at_match or not bl_match:
             raise Exception("Could not find session tokens. Are you logged in?")

        self.session_tokens = {
            "at": at_match.group(1),
            "bl": bl_match.group(1),
            "fsid": fsid_match.group(1) if fsid_match else ""
        }
        
        # Get cookies for requests
        self.cookies = await self.context.cookies()
        logger.info(f"[NotebookLM] ✅ Tokens acquired. bl: {self.session_tokens['bl']}")
        return True

    async def _execute_rpc(self, rpc_id: str, payload: Any) -> Any:
        # We can use the page to evaluate fetch, which automatically handles cookies and CORS best.
        # This is more robust than trying to reconstruct the request in python requests/aiohttp
        # because of the specific Google auth/cookies intricacies.
        
        req_id = random.randint(100000, 200000)
        inner_payload = json.dumps(payload)
        envelope = json.dumps([[[rpc_id, inner_payload, None, "generic"]]])
        
        params = {
            "rpcids": rpc_id,
            "source-path": "/",
            "bl": self.session_tokens["bl"],
            "f.sid": self.session_tokens["fsid"],
            "hl": "en",
            "rt": "c",
            "_reqid": str(req_id)
        }
        
        # Construct query string manually to ensure order if needed, but dict is usually fine
        query_string = "&".join([f"{k}={v}" for k, v in params.items()])
        url = f"{RPC_ENDPOINT}?{query_string}"

        # We need to send form data: f.req and at
        # We'll use page.evaluate to run fetch in the browser context
        
        js_code = """
        async ([url, envelope, at]) => {
            const body = new URLSearchParams();
            body.append("f.req", envelope);
            body.append("at", at);

            const response = await fetch(url, {
                method: "POST",
                headers: {
                    "Content-Type": "application/x-www-form-urlencoded;charset=UTF-8",
                    "X-Same-Domain": "1"
                },
                body: body,
            });
            
            if (!response.ok) throw new Error("RPC Failed: " + response.status);
            return await response.text();
        }
        """
        
        try:
            response_text = await self.page.evaluate(js_code, [url, envelope, self.session_tokens["at"]])
            return self._parse_rpc_response(response_text)
        except Exception as e:
            logger.error(f"[NotebookLM] RPC {rpc_id} failed: {e}")
            raise

    def _parse_rpc_response(self, text: str):
        lines = text.split('\n')
        for line in lines:
            trimmed = line.strip()
            if trimmed.startswith('[['):
                try:
                    data = json.loads(trimmed)
                    if data and data[0] and data[0][0] == 'wrb.fr':
                        return data
                except:
                    pass
        return None

    async def download_resource(self, url: str) -> bytes:
        """Downloads a resource (image) using the authenticated browser context."""
        logger.info(f"[NotebookLM] Downloading resource via Playwright API: {url[:50]}...")
        
        # Use Playwright's APIRequest context. 
        # This shares cookies with the browser but runs outside the page sandbox, avoiding CORS.
        if not self.context:
             await self.start()
             
        try:
            # We explicitly pass the referer to look legitimate
            response = await self.context.request.get(
                url, 
                headers={"Referer": "https://notebooklm.google.com/"}
            )
            
            if not response.ok:
                logger.error(f"[NotebookLM] Download failed: {response.status} {response.status_text}")
                # Try to log the body if it fails, might be a redirect or auth page
                try:
                    text = await response.text()
                    logger.error(f"Error Body: {text[:200]}")
                except:
                    pass
                raise Exception(f"Failed to download image: {response.status}")
                
            return await response.body()
            
        except Exception as e:
            logger.error(f"[NotebookLM] Browser download failed: {e}")
            raise

    def _find_source_id(self, obj: Any) -> Optional[str]:
        # UUID regex pattern
        uuid_pattern = re.compile(r'^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$', re.IGNORECASE)
        
        if isinstance(obj, str):
            if uuid_pattern.match(obj):
                return obj
        
        if isinstance(obj, list):
            for item in obj:
                found = self._find_source_id(item)
                if found: return found
        return None

    def _find_image_url(self, obj: Any) -> Optional[str]:
        if isinstance(obj, str):
            if 'googleusercontent.com' in obj or obj.startswith('data:image/'):
                return obj
        
        if isinstance(obj, list):
            for item in obj:
                found = self._find_image_url(item)
                if found: return found
        return None

    async def generate_infographic(self, video_url: str) -> str:
        if not self.session_tokens["at"]:
            await self.start()

        # --- CACHE CHECK ---
        cache_file = os.path.join(os.path.dirname(os.path.abspath(__file__)), "cache.json")
        cache = {}
        if os.path.exists(cache_file):
            try:
                with open(cache_file, "r") as f:
                    cache = json.load(f)
            except Exception as e:
                logger.warning(f"Failed to load cache: {e}")

        if video_url in cache:
            notebook_id = cache[video_url]
            logger.info(f"[NotebookLM] ⚡ Cache Hit! Reusing notebook: {notebook_id}")
            # We skip creation and source addition, just poll this notebook.
            return await self.poll_for_artifacts(notebook_id)

        # 1. Create Notebook
        logger.info("[NotebookLM] Creating Notebook...")
        create_payload = ["", None, None, [2], [1, None, None, None, None, None, None, None, None, None, [1]]]
        create_res = await self._execute_rpc(RPC_CREATE_NOTEBOOK, create_payload)
        
        inner_create = json.loads(create_res[0][2])
        notebook_id = inner_create[2]
        logger.info(f"[NotebookLM] Notebook Created: {notebook_id}")
        
        # Save state for recovery & cache
        try:
            # Update Last Run
            state_file = os.path.join(os.path.dirname(os.path.abspath(__file__)), "last_run.json")
            with open(state_file, "w") as f:
                json.dump({"last_notebook_id": notebook_id, "timestamp": time.time()}, f)
            
            # Update Cache
            cache[video_url] = notebook_id
            with open(cache_file, "w") as f:
                json.dump(cache, f)
                
        except Exception as e:
            logger.warning(f"Failed to save state/cache: {e}")

        # 2. Add Source
        logger.info(f"[NotebookLM] Adding Source: {video_url}...")
        source_payload = [[[None, None, None, None, None, None, None, [video_url], None, None, 1]], notebook_id, [2], [1, None, None, None, None, None, None, None, None, None, [1]]]
        source_res = await self._execute_rpc(RPC_ADD_SOURCE, source_payload)
        
        if not source_res or not source_res[0] or len(source_res[0]) < 3:
            logger.error(f"[NotebookLM] Invalid Add Source Response: {source_res}")
            raise Exception("Failed to add source: Invalid RPC response")

        raw_inner = source_res[0][2]
        if raw_inner is None:
             logger.error(f"[NotebookLM] Add Source Inner Payload is None. Full Res: {source_res}")
             raise Exception("Failed to add source: Google returned no data (Video might be rejected)")

        inner_source = json.loads(raw_inner)
        source_id = self._find_source_id(inner_source)
        
        if not source_id:
            raise Exception("Failed to add source. Google rejected the video (No transcript?).")
        
        logger.info(f"[NotebookLM] Source Added: {source_id}")

        # 3. Wait 
        logger.info("[NotebookLM] ⏳ Waiting 10s for transcript processing...")
        await asyncio.sleep(10)

        # 4. Trigger Generation
        logger.info("[NotebookLM] 🚀 Triggering Generation...")
        trigger_payload = [[2], notebook_id, [None, None, 7, [[[source_id]]], None, None, None, None, None, None, None, None, None, None, [[None, None, None, 1, 2]]]]
        await self._execute_rpc(RPC_GENERATE_INFOGRAPHIC, trigger_payload)
        
        # 5. Poll
        return await self.poll_for_artifacts(notebook_id)

    async def poll_for_artifacts(self, notebook_id: str) -> str:
        logger.info("[NotebookLM] Polling for artifacts...")
        for i in range(30):
            try:
                payload = [[2], notebook_id, 'NOT artifact.status = "ARTIFACT_STATUS_SUGGESTED"']
                response = await self._execute_rpc(RPC_LIST_ARTIFACTS, payload)
                
                if response and response[0] and isinstance(response[0][2], str):
                    inner_data = json.loads(response[0][2])
                    image_url = self._find_image_url(inner_data)
                    
                    if image_url:
                        logger.info(f"[NotebookLM] 📸 Image Found: {image_url}")
                        return image_url
            except Exception as e:
                logger.warning(f"[NotebookLM] Poll error: {e}")
            
            await asyncio.sleep(10)
            logger.info(f"[NotebookLM] Poll attempt {i+1}/30...")
            
        raise Exception("Timeout waiting for artifact generation")

# Example Usage
if __name__ == "__main__":
    async def main():
        # Set headless=False to login manually once
        client = NotebookLMClient(headless=False)
        try:
            await client.start()
            # Replace with a real video URL to test
            # img = await client.generate_infographic("https://www.youtube.com/watch?v=YOUR_VIDEO_ID")
            # print("Final Image:", img)
        finally:
            # await client.stop() # Keep open to see result if testing
            pass
            
    # asyncio.run(main())

File: src/notebooklm_client.ts

import { chromium, BrowserContext, Page, APIRequestContext } from 'playwright';
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';

// --- CONFIGURATION ---
const BASE_URL = "https://notebooklm.google.com";
const RPC_ENDPOINT = `${BASE_URL}/_/LabsTailwindUi/data/batchexecute`;
const RPC_GENERATE_STREAMED = `${BASE_URL}/_/LabsTailwindUi/data/google.internal.labs.tailwind.orchestration.v1.LabsTailwindOrchestrationService/GenerateFreeFormStreamed`;

// RPC IDs
const RPC_CREATE_NOTEBOOK = "CCqFvf";
const RPC_ADD_SOURCE = "izAoDd";
const RPC_GENERATE_INFOGRAPHIC = "R7cb6c";
const RPC_LIST_ARTIFACTS = "gArtLc";
const RPC_DELETE_NOTEBOOK = "f61S6e";

interface SessionTokens {
    at: string | null;
    bl: string | null;
    fsid: string | null;
}

export class NotebookLMClient {
    private headless: boolean;
    private context: BrowserContext | null = null;
    private page: Page | null = null;
    private sessionTokens: SessionTokens = { at: null, bl: null, fsid: null };

    constructor(headless: boolean = true) {
        this.headless = headless;
    }

    async start(): Promise<void> {
        // use /tmp to avoid CWD/Permission issues in Claude Desktop
        const userDataDir = "/tmp/notebooklm_user_data";

        console.error(`[NotebookLM] Launching browser (Headless: ${this.headless})...`);
        this.context = await chromium.launchPersistentContext(userDataDir, {
            headless: this.headless,
            args: [
                "--disable-blink-features=AutomationControlled",
                "--ignore-certificate-errors",
                "--no-sandbox",
                "--disable-setuid-sandbox"
            ]
        });

        this.page = this.context.pages()[0] || await this.context.newPage();
        await this._refreshTokens();
    }

    async stop(): Promise<void> {
        if (this.context) {
            await this.context.close();
        }
    }

    async openLoginWindow(): Promise<void> {
        console.error("[NotebookLM] 🔓 Launching Interactive Login Window...");
        await this.stop(); // Release lock

        // Temporarily act as not-headless
        this.headless = false;
        await this.start();

        if (this.page) {
            await this.page.goto("https://notebooklm.google.com");
            console.error("[NotebookLM] Login Window Opened. Waiting for user to log in...");

            try {
                // Wait for the 'Create new' button which indicates we are on the dashboard (logged in)
                await this.page.waitForSelector('text="Create new"', { timeout: 300000 });
                console.error("[NotebookLM] ✅ Login detected! Closing window in 2 seconds...");
                await this.page.waitForTimeout(2000);
            } catch (error) {
                console.error("[NotebookLM] ⚠️ Login verification timed out or failed. Closing window.");
            }

            await this.stop();
            // Reset to headless for next time
            this.headless = true;
        }
    }

    async _refreshTokens(): Promise<boolean> {
        console.error("[NotebookLM] 🔄 Navigating to scrape tokens...");
        if (!this.page) throw new Error("Page not initialized");

        await this.page.goto(BASE_URL);

        if (this.page.url().includes("accounts.google.com")) {
            console.warn("[NotebookLM] ⚠️ Login required!");
            if (this.headless) {
                // Return specific error that server.ts can catch
                throw new Error("Authentication required. Please run with headless=false first to login.");
            }
            // Wait for user login
            await this.page.waitForURL("https://notebooklm.google.com/**", { timeout: 0 });
            console.error("[NotebookLM] Login detected.");
        }

        let content = await this.page.content();

        // Regex extraction
        const atMatch = content.match(/"SNlM0e":"([^"]+)"/);
        const blMatch = content.match(/"(boq_labs-tailwind-[^"]+)"/);
        const fsidMatch = content.match(/"FdrFJe":"([^"]+)"/);

        if (!atMatch || !blMatch) {
            await this.page.waitForTimeout(2000);
            content = await this.page.content();
        }

        const atV = content.match(/"SNlM0e":"([^"]+)"/)?.[1];
        const blV = content.match(/"(boq_labs-tailwind-[^"]+)"/)?.[1];
        const fsidV = content.match(/"FdrFJe":"([^"]+)"/)?.[1];

        if (!atV || !blV) {
            throw new Error("Could not find session tokens. Are you logged in?");
        }

        this.sessionTokens = {
            at: atV,
            bl: blV,
            fsid: fsidV || null
        };

        console.error(`[NotebookLM] ✅ Tokens acquired. bl: ${this.sessionTokens.bl}`);
        return true;
    }

    async _executeRpc(rpcId: string, payload: any): Promise<any> {
        if (!this.page) throw new Error("Client not started");

        const reqId = Math.floor(Math.random() * 100000) + 100000;
        const innerPayload = JSON.stringify(payload);
        const envelope = JSON.stringify([[[rpcId, innerPayload, null, "generic"]]]);

        const params = new URLSearchParams({
            "rpcids": rpcId,
            "source-path": "/",
            "bl": this.sessionTokens.bl || "",
            "f.sid": this.sessionTokens.fsid || "",
            "hl": "en",
            "rt": "c",
            "_reqid": reqId.toString()
        });

        const url = `${RPC_ENDPOINT}?${params.toString()}`;

        // Executing fetch inside the page context
        const responseText = await this.page.evaluate(async ({ url, envelope, at }) => {
            const body = new URLSearchParams();
            body.append("f.req", envelope);
            body.append("at", at);

            const res = await fetch(url, {
                method: "POST",
                headers: {
                    "Content-Type": "application/x-www-form-urlencoded;charset=UTF-8",
                    "X-Same-Domain": "1"
                },
                body: body
            });
            if (!res.ok) throw new Error("RPC Failed: " + res.status);
            return await res.text();
        }, { url, envelope, at: this.sessionTokens.at || "" });

        return this._parseRpcResponse(responseText);
    }

    _parseRpcResponse(text: string): any {
        const lines = text.split('\n');
        for (const line of lines) {
            const trimmed = line.trim();
            if (trimmed.startsWith('[[')) {
                try {
                    const data = JSON.parse(trimmed);
                    if (data && data[0] && data[0][0] === 'wrb.fr') {
                        return data;
                    }
                } catch (e) { }
            }
        }
        return null;
    }

    async _executeStreamedRpc(fReqPayload: any): Promise<Buffer> {
        if (!this.context) throw new Error("Client not started");

        const reqId = Math.floor(Math.random() * 100000) + 100000;
        const fReqStr = JSON.stringify(fReqPayload); // separators not critical in JS usually, but standard JSON is fine

        const params = new URLSearchParams({
            "bl": this.sessionTokens.bl || "",
            "f.sid": this.sessionTokens.fsid || "",
            "hl": "en",
            "_reqid": reqId.toString(),
            "rt": "c"
        });

        const url = `${RPC_GENERATE_STREAMED}?${params.toString()}`;
        console.error(`[NotebookLM] Executing Streamed RPC to ${url}`);

        const response = await this.context.request.post(url, {
            form: {
                "f.req": fReqStr,
                "at": this.sessionTokens.at || ""
            },
            headers: {
                "X-Same-Domain": "1"
            },
            timeout: 120000
        });

        if (!response.ok()) {
            throw new Error(`Streamed RPC Failed: ${response.status()} ${await response.text()}`);
        }

        return await response.body();
    }

    _parseStreamedResponse(entryBuffer: Buffer): string {
        let fullText = "";
        let textBody = entryBuffer.toString('utf-8');

        // Remove XSSI guard
        if (textBody.startsWith(")]}'")) {
            textBody = textBody.substring(4).trim();
        }

        // Robust JSON Scan
        // In JS we don't have raw_decode returning end_pos. We must find the bracket balance.
        let pos = 0;
        const len = textBody.length;

        while (pos < len) {
            // finding start '['
            const startBracket = textBody.indexOf('[', pos);
            if (startBracket === -1) break;

            // Find balanced ending ']'
            const endBracket = this._findBalancedEnd(textBody, startBracket);
            if (endBracket === -1) {
                // Incomplete JSON or malformed
                pos = startBracket + 1;
                continue;
            }

            const jsonStr = textBody.substring(startBracket, endBracket + 1);
            try {
                const obj = JSON.parse(jsonStr);
                const extracted = this._extractWrbText(obj);
                if (extracted) {
                    // Last Write Wins
                    fullText = extracted.trim() + "\n";
                }
            } catch (e) {
                // Ignore parse errors
            }

            pos = endBracket + 1;
        }

        return fullText.trim();
    }

    // Helper for finding the matching closing bracket
    _findBalancedEnd(str: string, start: number): number {
        let depth = 0;
        let inString = false;
        let escape = false;

        for (let i = start; i < str.length; i++) {
            const char = str[i];

            if (escape) {
                escape = false;
                continue;
            }

            if (char === '\\') {
                escape = true;
                continue;
            }

            if (char === '"') {
                inString = !inString;
                continue;
            }

            if (!inString) {
                if (char === '[') depth++;
                else if (char === ']') {
                    depth--;
                    if (depth === 0) return i;
                }
                else if (char === '{') depth++; // Should handle object braces too just in case
                else if (char === '}') depth--;
            }
        }
        return -1;
    }

    _extractWrbText(node: any): string {
        const results: string[] = [];

        const walk = (n: any, inPayload: boolean = false) => {
            if (Array.isArray(n)) {
                // wrb.fr check
                if (n.length >= 3 && n[0] === "wrb.fr" && typeof n[2] === 'string') {
                    try {
                        const innerJson = n[2];
                        if (innerJson.trim().startsWith("[")) {
                            const decoded = JSON.parse(innerJson);
                            if (Array.isArray(decoded) && decoded.length > 2) {
                                walk(decoded[0], true);
                            }
                        }
                    } catch (e) { }
                } else if (inPayload) {
                    // Anti-Transcript Heuristic
                    // Check for [start(int), end(int), ...] or [null, start(int), end(int)]
                    // JS numbers are just numbers.

                    if (n.length >= 2 && typeof n[0] === 'number' && typeof n[1] === 'number') return;
                    if (n.length >= 3 && n[0] === null && typeof n[1] === 'number' && typeof n[2] === 'number') return;

                    n.forEach(c => walk(c, inPayload));
                } else {
                    n.forEach(c => walk(c, inPayload));
                }
            } else if (typeof n === 'string' && inPayload) {
                const val = n.trim();
                // UUID Filter & Empty
                if (val && val.length !== 36) {
                    results.push(val);
                }
            }
        };

        walk(node);
        return results.join("\n");
    }

    async downloadResource(url: string): Promise<Buffer> {
        console.error(`[NotebookLM] Downloading resource: ${url.substring(0, 50)}...`);
        if (!this.context) await this.start();
        if (!this.context) throw new Error("Context failed to start");

        const response = await this.context.request.get(url, {
            headers: { "Referer": "https://notebooklm.google.com/" }
        });

        if (!response.ok()) {
            throw new Error(`Failed to download image: ${response.status()}`);
        }
        return await response.body();
    }

    _findImageUrl(obj: any): string | null {
        if (typeof obj === 'string') {
            if (obj.includes('googleusercontent.com') || obj.startsWith('data:image/')) return obj;
        }
        if (Array.isArray(obj)) {
            for (const item of obj) {
                const found = this._findImageUrl(item);
                if (found) return found;
            }
        }
        return null;
    }

    _findSourceId(obj: any): string | null {
        // rough uuid pattern check
        const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
        if (typeof obj === 'string') {
            if (uuidRegex.test(obj)) return obj;
        }
        if (Array.isArray(obj)) {
            for (const item of obj) {
                const found = this._findSourceId(item);
                if (found) return found;
            }
        }
        return null;
    }

    async prepareNotebook(videoUrl: string): Promise<{ notebookId: string, sourceId: string }> {
        if (!this.sessionTokens.at) await this.start();

        const cacheFile = "/Users/harivishnus/Desktop/University/Internships/altrosyn/Chrome_extension/cache.json";
        let cache: any = {};
        if (fs.existsSync(cacheFile)) {
            try {
                cache = JSON.parse(fs.readFileSync(cacheFile, 'utf-8'));
            } catch (e) { }
        }

        let notebookId: string | null = null;
        let sourceId: string | null = null;

        if (cache[videoUrl]) {
            const entry = cache[videoUrl];
            if (typeof entry === 'object') {
                notebookId = entry.notebookId || entry.notebook_id; // handle both
                sourceId = entry.sourceId || entry.source_id;
            } else {
                notebookId = entry;
            }
            if (notebookId) console.error(`[NotebookLM] ⚡ Cache Hit! Reusing notebook: ${notebookId}`);
        }

        if (!notebookId) {
            console.error("[NotebookLM] Creating Notebook...");
            const createPayload = ["", null, null, [2], [1, null, null, null, null, null, null, null, null, null, [1]]];
            const createRes = await this._executeRpc(RPC_CREATE_NOTEBOOK, createPayload);
            const innerCreate = JSON.parse(createRes[0][2]);
            notebookId = innerCreate[2];
            console.error(`[NotebookLM] Notebook Created: ${notebookId}`);
        }

        if (!sourceId) {
            console.error(`[NotebookLM] Adding Source: ${videoUrl}...`);
            // NOTE: Python used 'None' which became 'null' in JSON. JS 'null' matches.
            const sourcePayload = [[[null, null, null, null, null, null, null, [videoUrl], null, null, 1]], notebookId, [2], [1, null, null, null, null, null, null, null, null, null, [1]]];
            const sourceRes = await this._executeRpc(RPC_ADD_SOURCE, sourcePayload);

            if (!sourceRes || !sourceRes[0]) throw new Error("Invalid Add Source Response");

            const rawInner = sourceRes[0][2];
            const innerSource = JSON.parse(rawInner);
            sourceId = this._findSourceId(innerSource);

            if (!sourceId) throw new Error("Failed to add source (No ID found)");

            console.error(`[NotebookLM] Source Added: ${sourceId}`);

            cache[videoUrl] = { notebookId, sourceId };
            fs.writeFileSync(cacheFile, JSON.stringify(cache, null, 2));
        } else {
            console.error(`[NotebookLM] ⚡ Using Cached Source ID: ${sourceId}`);
        }

        return { notebookId: notebookId!, sourceId: sourceId! };
    }

    async pollForArtifacts(notebookId: string): Promise<string> {
        console.error("[NotebookLM] Polling for artifacts...");
        for (let i = 0; i < 30; i++) {
            try {
                const payload = [[2], notebookId, 'NOT artifact.status = "ARTIFACT_STATUS_SUGGESTED"'];
                const response = await this._executeRpc(RPC_LIST_ARTIFACTS, payload);
                if (response && response[0] && typeof response[0][2] === 'string') {
                    const innerData = JSON.parse(response[0][2]);
                    const imageUrl = this._findImageUrl(innerData);
                    if (imageUrl) {
                        console.error(`[NotebookLM] 📸 Image Found: ${imageUrl}`);
                        return imageUrl;
                    }
                }
            } catch (e) { }
            await new Promise(r => setTimeout(r, 10000));
            console.error(`[NotebookLM] Poll attempt ${i + 1}/30...`);
        }
        throw new Error("Timeout waiting for artifact creation");
    }

    async generateInfographic(videoUrl: string): Promise<string> {
        const { notebookId, sourceId } = await this.prepareNotebook(videoUrl);
        console.error("[NotebookLM] ⏳ Waiting 5s...");
        await new Promise(r => setTimeout(r, 5000));

        console.error("[NotebookLM] 🚀 Triggering Infographic...");
        const triggerPayload = [[2], notebookId, [null, null, 7, [[[sourceId]]], null, null, null, null, null, null, null, null, null, null, [[null, null, null, 1, 2]]]];
        await this._executeRpc(RPC_GENERATE_INFOGRAPHIC, triggerPayload);

        return await this.pollForArtifacts(notebookId);
    }

    async queryNotebook(notebookId: string, sourceId: string, prompt: string): Promise<string> {
        console.error(`[NotebookLM] Parsing query: "${prompt}" for notebook ${notebookId}...`);

        const innerReq = [
            [[[sourceId]]],
            prompt,
            null,
            [2, null, [1], [1]],
            null, null, null,
            notebookId,
            1
        ];

        const fReq = [
            null,
            JSON.stringify(innerReq)
        ];

        const rawResponse = await this._executeStreamedRpc(fReq);
        const summary = this._parseStreamedResponse(rawResponse);

        if (!summary) {
            console.warn("[NotebookLM] Query returned empty text.");
            return "Failed to generate answer.";
        }

        console.error("[NotebookLM] Query successful.");
        return summary;
    }

    async query(videoUrl: string, question: string): Promise<string> {
        const { notebookId, sourceId } = await this.prepareNotebook(videoUrl);

        // Small delay to ensure state consistency
        console.error("[NotebookLM] ⏳ Waiting 2s before querying...");
        await new Promise(r => setTimeout(r, 2000));

        return await this.queryNotebook(notebookId, sourceId, question);
    }

    async generateSummary(videoUrl: string): Promise<string> {
        // Wrapper for backward compatibility and specific "Summary" behavior (longer wait)
        const { notebookId, sourceId } = await this.prepareNotebook(videoUrl);

        console.error("[NotebookLM] ⏳ Waiting 10s before requesting summary...");
        await new Promise(r => setTimeout(r, 10000));

        return await this.queryNotebook(notebookId, sourceId, "give me summary of the video");
    }
}
